### PR TITLE
Generate previews for songs

### PR DESCRIPTION
**Pull request by bui**
This lowers the filesize and thus loading times for previews.

taiko-web now requires ffmpeg and `ffmpy` python module to be installed.

An nginx conf update is also required to redirect missing previews to the api server:

```
location ~ ^/(assets|songs|src)/ {
    root   /home/bui/taiko-web/public;
    location ~ ^/songs/([0-9]+)/preview\.mp3$ {
        try_files $uri /api/preview?id=$1;
    }
}
```

### CODE REVIEW

File: public/src/js/songselect.js [hunk 1]
```diff
@@ -1285,7 +1286,7 @@ class SongSelect{
 		var difference = endLoad - startLoad
 		var minDelay = 300
 		var delay = minDelay - Math.min(minDelay, difference)
-		this.preview.playLoop(delay / 1000, false, prvtime / 1000)
+		this.preview.playLoop(delay / 1000, false, 0)
```
Comment (LoveEevee) [inline public/src/js/songselect.js]:
```suggestion
		this.preview.playLoop(delay / 1000, false, prvtime > 0.1 ? 0 : prvtime)
```

Comment (bui) [reply public/src/js/songselect.js]: Can you explain this? We always want to start the preview file from the start (it is already cropped)

**Comment by LoveEevee** [comment]
Can the files be compressed even more, like lower the bitrate and making them mono
Half of the file size does not change preview load times by much

File: app.py [hunk 2]
```diff
@@ -138,5 +163,34 @@ def route_api_songs():
...
+
+    if os.path.isfile(song_path) and not os.path.isfile(prev_path):
+        preview = get_preview(song_id, song_type) / 1000
+        if not preview or preview <= 0.1:
+            print 'Skipping #%s due to no preview' % song_id
+            return False
+
+        print 'Making preview.mp3 for song #%s' % song_id
+        ff = FFmpeg(inputs={song_path: '-ss %s' % preview},
+                    outputs={prev_path: '-codec:a libmp3lame -b:a 128k -y -loglevel panic'})
```
Comment (LoveEevee) [inline app.py]:
```suggestion
                    outputs={prev_path: '-codec:a libmp3lame -ar 32000 -b:a 92k -y -loglevel panic'})
```

**Comment by LoveEevee** [comment]
Merge this commit pls
https://github.com/LoveEevee/taiko-web/commit/a4ba1009fd3b1c441ac7d0e1caeb629767229c40

**Comment by bui** [comment]
push it to my branch

**Comment by bui** [comment]
I don't understand what you're trying to do here

The intended behaviour is:
- if a song has a preview time defined (by api), it will fetch preview.mp3 and play from the start
- if there is no preview defined  (or preview <= 0.1 -- to account for charts with preview set to something like 0.01), it will load main.mp3 instead and also play it from the start

Making a request to preview.mp3 when it might not exist is unnecessary

**Comment by LoveEevee** [comment]
```js
loadPreview(previewFilename).catch(() => {
```
This code will try to load an mp3 file and if it does not exit or there is an error it will try to do something else. There are many reasons to do it:

1. When loading preview.mp3 for the first time it is not guaranteed to generate correctly, ffmpeg may exit without generating the file
2. The conversion can take longer than the timeout of webserver
3. You may run taiko-web without app.py running and there are no previews

There should be a way to fallback to main.mp3 with the offset in those cases and my commit does exactly that. `songObj.preview_time` is a variable that defines the music offset, which is 0 until there is an error getting preview.mp3

**Comment by bui** [comment]
OK. Please push your changes to my branch so I can merge them

**Pull request merged**